File: ParserImportacao.java
import java.io.*;

public class ParserImportacao {

   private Empresa empresa;

   ParserImportacao(Empresa empresa) {
      this.empresa = empresa;
   }

   // Importa dados do Arquivo produtos.txt
   void importaEstoque(String NomeArquivo) {
      String linha = null;
      try (BufferedReader leitor = new BufferedReader(new FileReader(NomeArquivo))) {
         // 16320;344;200
         // 16320 -> Código do Produto;
         // 344 -> Quantidade em estoque no início do período;
         // 200 -> Quantidade mínima que deve ser mantida no Centro Operacional
         while ((linha = leitor.readLine()) != null) {
            String[] campos = linha.split(";");
            int codigo = Integer.parseInt(campos[0]);
            int qtInicial = Integer.parseInt(campos[1]);
            int qtCO = Integer.parseInt(campos[2]);
            empresa.registraProduto(codigo, qtInicial, qtCO);
         }
      } catch (FileNotFoundException e) { e.printStackTrace();
      } catch (IOException e) { e.printStackTrace(); }
   }

   // Importa dados do Arquivo vendas.txt
   // Importa dados para formar o arquivo DIVERGENCIAS.TXT
   // Importa dados para formar o arquivo TOTCANAIS.TXT
   void importaVenda(String NomeArquivo) {
      int contadorLinhas = 0;
      String linha = null;
      try (BufferedReader leitor = new BufferedReader(new FileReader(NomeArquivo))) {
         // 16320;1;100;2
         // 16320 -> Código do Produto
         // 1 -> Quantidade vendida
         // 100 -> Situação da venda
         // 2 -> Canal de venda
         while ((linha = leitor.readLine()) != null) {
            String[] campos = linha.split(";");
            contadorLinhas++;
            int codigo = Integer.parseInt(campos[0]);
            int qtVendida = Integer.parseInt(campos[1]);
            int situacaoVenda = Integer.parseInt(campos[2]);
            int canalVenda = Integer.parseInt(campos[3]);

            if (situacaoVenda == 100 || situacaoVenda == 102) {
               if (empresa.contemEstoque(codigo)) {
                  if (empresa.contemVenda(codigo))
                     empresa.incrementaVenda(codigo, qtVendida);
                  else
                     empresa.registraVenda(codigo, qtVendida, situacaoVenda, canalVenda);
               }
            }

            //  Relatório de Divergências
            // Linha LL – Código de Produto não encontrado CCCCC
            if (!empresa.contemEstoque(codigo) && situacaoVenda != 999) {
               StringBuilder s = new StringBuilder();
               s.append("Linha " + contadorLinhas + " - ");
               s.append("Código de Produto não encontrado ");
               s.append(codigo);
               empresa.registraDivergencia(s.toString());
            }

            StringBuilder s = new StringBuilder();
            switch (situacaoVenda) {
               case 135: // Linha LL – Venda cancelada
                  s.append("Linha " + contadorLinhas + " - Venda cancelada");
                  empresa.registraDivergencia(s.toString());
                  break;
               case 190: // Linha LL – Venda não finalizada
                  s.append("Linha " + contadorLinhas + " - Venda não finalizada");
                  empresa.registraDivergencia(s.toString());
                  break;
               case 999: // Linha LL – Erro desconhecido.
                  s.append("Linha " + contadorLinhas + " - ");
                  s.append("Erro desconhecido. Acionar equipe de TI");
                  empresa.registraDivergencia(s.toString());
                  break;
            }

            // Relatório por canais
            if (situacaoVenda == 100 || situacaoVenda == 102) {
               switch (canalVenda) {
                  case 1: // 1: Representante comercial
                     empresa.incrementaRepresentantes(qtVendida);
                     break;
                  case 2: // 2: Website
                     empresa.incrementaWebSite(qtVendida);
                     break;
                  case 3: // 3: Aplicativo móvel Android
                     empresa.incrementaAppAndroid(qtVendida);
                     break;
                  case 4: // 4: Aplicativo móvel iPhone
                     empresa.incrementaAppIphone(qtVendida);
                     break;
               }
            }
         }
      } catch (FileNotFoundException e) { e.printStackTrace();
      } catch (IOException e) { e.printStackTrace(); }
   }
}

File: parser_importacao.c
/*
 * File:  parser_importacao.c
 * Autor:  Laercio Costa Rodrigues
 * Descricao: Teste Técnico Intelitrader
*/

#include "parser_importacao.h"


/* Importa dados do Arquivo produtos.txt / retorna lista_estoque */
/* 16320;344;200 */
/* 16320 -> Código do Produto; */
/* 344 -> Quantidade em estoque no início do período; */
/* 200 -> Quantidade mínima que deve ser mantida no Centro Operacional */
lista_estoque* importa_estoque(char* arquivo_vendas) {

   int codigo, qtInicial, qtCO;
   lista_estoque* pt_lista_estoque = cria_lista_estoque();
   FILE *leitor;
   leitor = fopen(arquivo_vendas,"r");

   while(fscanf(leitor, "%d;%d;%d", &codigo, &qtInicial, &qtCO) != EOF) {
      registra_produto(pt_lista_estoque, codigo, qtInicial, qtCO);
   }

   fclose(leitor);
   return pt_lista_estoque;
}


/* Importa dados do Arquivo vendas.txt */
/* Exporta arquivo DIVERGENCIAS.TXT*/
/* Retorna lista venda */
/* 16320;1;100;2 */
/* 16320 -> Código do Produto */
/* 1 -> Quantidade vendida */
/* 100 -> Situação da venda */
/* 2 -> Canal de venda */
lista_venda* importa_venda(char* arquivo_vendas, char* arquivo_divergencias, 
                                           lista_estoque* pt_lista_estoque) {

   int codigo, qtVendida, situacaoVenda, canalVenda, contador_linhas = 0;
   lista_venda* pt_lista_venda = cria_lista_venda();
   FILE *leitor;
   FILE* escritor;
   leitor = fopen(arquivo_vendas,"r");
   /* Criar novo arquivo texto se ja existir um antigo */
   escritor = fopen(arquivo_divergencias, "w");
   fclose(escritor);   

   /* Inicializa valores de vendas por canal */
   inicializa_vendas_por_canal();

   while(fscanf(leitor, "%d;%d;%d;%d", &codigo, &qtVendida, &situacaoVenda, &canalVenda) != EOF) {

      contador_linhas++;

      if (situacaoVenda == 100 || situacaoVenda == 102) {

         if (procura_no_estoque(pt_lista_estoque, codigo) != NULL) {

            if (procura_no_venda(pt_lista_venda, codigo) != NULL)
               incrementa_venda(pt_lista_venda, codigo, qtVendida);
            else {
               registra_venda(pt_lista_venda, codigo, qtVendida, situacaoVenda, canalVenda);
            }
         }
      }
               
      /* Relatório de Divergências */
      /* Linha LL – Código de Produto não encontrado CCCCC */
      if (procura_no_estoque(pt_lista_estoque, codigo) == NULL && situacaoVenda != 999) {
         divergencia_nao_encontrado(contador_linhas, codigo, arquivo_divergencias);
      }

      switch (situacaoVenda) {
         case 135: /* Linha LL – Venda cancelada */
               divergencia_cancelada(contador_linhas, arquivo_divergencias);
               break;
         case 190: /* Linha LL – Venda não finalizada */
               divergencia_nao_finalizada(contador_linhas, arquivo_divergencias);
               break;
         case 999: /* Linha LL – Erro desconhecido. */
               divergencia_erro(contador_linhas, arquivo_divergencias);
               break;
      }

      /* Relatório por canais */
      if (situacaoVenda == 100 || situacaoVenda == 102) {
         switch (canalVenda) {
            case 1:
               incrementa_representantes(qtVendida);
               break;
            case 2:
               incrementa_website(qtVendida);
               break;
            case 3:
               incrementa_app_android(qtVendida);
               break;
            case 4:
               incrementa_app_iphone(qtVendida);
               break;
         }
      }

   }

   fclose(leitor);
   return pt_lista_venda;
}